node/lib/util/stash_util.js

```diff
@@ -148,30 +151,159 @@ exports.save = co.wrap(function *(repo, status, includeUntracked, message) {
-             (!includeUntracked ||
-              0 === Object.keys(wd.status.workdir).length))) {
-            // Nothing to do for closed or clean subs
-
-            return;                                                   // RETURN
-        }
-        const subRepo = yield SubmoduleUtil.getRepo(repo, name);
-        subRepos[name] = subRepo;
 
         let stashId;
-        if (sub.commit !== null && wd.status.headCommit === sub.commit.sha) {
-            const FLAGS = NodeGit.Stash.FLAGS;
-            const flags = includeUntracked ?
-                  FLAGS.INCLUDE_UNTRACKED :
-                  FLAGS.DEFAULT;
-            stashId = yield NodeGit.Stash.save(subRepo, sig, "stash",
-                                                     flags);
+
+        if (sub.commit === null) {
+            // I genuinely have no idea when this happens -- it's not:
+            // (a) a closed submodule with a change staged in the meta repo
+            // (b) a submodule yet to be born -- that is, a submodule
+            // added to .gitmodules but without any commits.
+            // (c) a submodule which does not even appear inside the
+            // .gitmodules (e.g. one that you meant to add but didn't)
+            console.error(`BUG: ${name} is in an unexpected state. Please \
+report this.  Continuing stash anyway.`);
+            return;
         }
-        else {
-            stashId = NodeGit.Oid.fromString(wd.status.headCommit);
+
+        if (null === wd) {
+            // closed submodule
+            if (sub.commit.sha === sub.index.sha) {
+                // ... with no staged changes
+                return;                                               // RETURN
+            }
+            // This is a case that regular git stash doesn't really have
+            // to handle. In a normal stash commit, the tree points
+            // to the working directory tree, but here, there is no working
+            // directory.  But if there were, we would want to have
+            // this commit checked out.
+
+            const subRepo = yield SubmoduleUtil.getRepo(repo, name);
+
+            const subCommit = yield Commit.lookup(subRepo, sub.commit.sha);
+            const indexCommit = yield Commit.lookup(subRepo, sub.index.sha);
+            const indexTree = yield indexCommit.getTree();
+            stashId = yield Commit.create(subRepo,
+                                          null,
+                                          sig,
+                                          sig,
+                                          null,
+                                          "stash",
+                                          indexTree,
+                                          4,
+                                          [subCommit,
+                                           indexCommit,
+                                           indexCommit,
+                                           indexCommit]);
```

Comment:
why does this commit have 4 parents and three of them are the same?

Reply:
I'll edit the docs.